### ISSUE TITLE
[proposal] Want dry-struct compiler

### ISSUE THREAD

**Issue by YukiJikumaru** [Shopify/tapioca#1210]
I'm using [dry-rb](https://dry-rb.org/) which defining typed struct classes, authored by @solnic.

```
class User < Dry::Struct
  attribute :name, Types::String.optional
  attribute :age, Types::Coercible::Integer
end

class Location < Dry::Struct::Value
  attribute :lat, Types::Float
  attribute :lng, Types::Float
end
```

Nowadays  we can use [T::Struct](https://sorbet.org/docs/tstruct) to define typed struct classes, but my project have massive dry-rb structs and T::Struct is not fully compatible with dry-rb.

Therefore I need rbi files for dry-structs, and I wrote a dsl compiler `lib/tapioca/dsl/compilers/dry_struct.rb` .

[YukiJikumaru/tapioca](https://github.com/YukiJikumaru/tapioca/pull/1)

I wan to know 2 points.
A new dsl compiler is wellcome to tapioca? And you have any contribution guidelines?

**Comment by paracycle** [comment]
Thanks for writing the compiler for `dry-structs`, it looks great!

We only include DSL compilers for a few core gems that we use at Shopify inside Tapioca at this time. Our plan is to extract them, when appropriate, to the gems that they are modelling at some point in the future.

For that reason, we are not accepting DSL compiler contributions, since we won't be able to support them.

However, Tapioca is smart enough to load any Ruby file under the `lib/tapioca/dsl/compilers/` folder of any gem, so you can package that up as a gem, if you'd like. Moreover, if you think the dry-rb project would accept it as a contribution, you could send it upstream as well.

Finally, you are correct, we should make this more clear in our contribution guidelines.

**Comment by iMacTia** [comment]
@YukiJikumaru I like the idea of adding the compiler to `dry-rb`, this way we won't need to add extra gems to our Gemfile _just_ for compilers. Do you think it's worth opening a ticket with `dry-rb` and asking them if they'd be open to it?

**Comment by iMacTia** [comment]
That is very fair @solnic and as an OSS maintainer myself I can see your angle. I'd probably need to do something similar for Faraday as well as we do use quite a lot of metaprogramming 😅.
OTOH, I worked enough with typescript in the past to completely hate having to add all the `@types/my-labrary` to my projects, and I'd really love to spare this pain to Ruby engineers.

Bundling tapioca compilers inside gems will avoid the situation where `Gemfile`s gets full of `tapioca-*` gems that developers need to add manually. Instead, RBI support would become a plug-and-play thing.
The other advantage is that the compiler will likely need updating when the gem is updated, or it might break on some changes. Having it part of the gem itself would allow spotting these issues upstream with the right unit tests.

I do see the issue of the maintenance burden though, and I don't really have a good solution for that in mind.
It would be interesting to know if @paracycle and the rest of the team have already thought about this kind of integration and the potential problem of where the liability/maintainance burden falls wrt the compilers.

Finally, a super-wild thought: an alternative solution would be to have separate gems with some sort of convention (e.g. `tapioca-*` as suggested above), but have some sort of support in `bundler` that automatically pulls these extensions if you're working with tapioca/RBIs.

**Comment by YukiJikumaru** [comment]
@solnic 

Thank you! After reconsidering, I thought `tapioca-compilers-dry_struct`  should be better.

As recommendations for naming gems by https://guides.rubygems.org/name-your-gem/ .

GEM NAME | REQUIRE STATEMENT | MAIN CLASS OR MODULE
-- | -- | --
ruby_parser | require 'ruby_parser' | RubyParser
rdoc-data | require 'rdoc/data' | RDoc::Data
net-http-persistent | require 'net/http/persistent' | Net::HTTP::Persistent
net-http-digest_auth | require 'net/http/digest_auth' | Net::HTTP::DigestAuth
tapioca-compilers-dry_struct | require 'tapioca/compilers/dry_struct' | Tapioca::Compilers::DryStruct
tapioca-compilers-virtus | require 'tapioca/compilers/virtus' | Tapioca::Compilers::Virtus

**Comment by YukiJikumaru** [comment]
There is still a lot of discussion and there may be major changes, I have released the alpha version.

https://rubygems.org/gems/tapioca-compilers-dry_struct

**Comment by solnic** [comment]
Yeah I can see that now. Thanks for sharing your thoughts and insights. I'll talk to the rest of the team about this after we're done with Hanami 2.0 release. Supporting Sorbet OOTB would be **really nice**. I just want to be sure that if we decide to do it as part of the gem codebases, then it will actually be maintained 😅
